arch/arm/soc/nxp_imx/mcimx6x_m4/soc.c

```diff
@@ -0,0 +1,160 @@
+#endif /* CONFIG_UART_IMX_UART_2 */
+#ifdef CONFIG_UART_IMX_UART_3
+	/* Set access to UART_3 for M4 core */
+	RDC_SetPdapAccess(RDC, rdcPdapUart3,
+			RDC_DOMAIN_PERM(CONFIG_DOMAIN_ID, RDC_DOMAIN_PERM_RW),
+			false, false);
+#endif /* CONFIG_UART_IMX_UART_3 */
+#ifdef CONFIG_UART_IMX_UART_4
+	/* Set access to UART_4 for M4 core */
+	RDC_SetPdapAccess(RDC, rdcPdapUart4,
+			RDC_DOMAIN_PERM(CONFIG_DOMAIN_ID, RDC_DOMAIN_PERM_RW),
+			false, false);
+#endif /* CONFIG_UART_IMX_UART_4 */
+#ifdef CONFIG_UART_IMX_UART_5
+	/* Set access to UART_5 for M4 core */
+	RDC_SetPdapAccess(RDC, rdcPdapUart5,
+			RDC_DOMAIN_PERM(CONFIG_DOMAIN_ID, RDC_DOMAIN_PERM_RW),
+			false, false);
+#endif /* CONFIG_UART_IMX_UART_5 */
+#ifdef CONFIG_UART_IMX_UART_6
+	/* Set access to UART_6 for M4 core */
+	RDC_SetPdapAccess(RDC, rdcPdapUart6,
+			RDC_DOMAIN_PERM(CONFIG_DOMAIN_ID, RDC_DOMAIN_PERM_RW),
+			false, false);
+#endif /* CONFIG_UART_IMX_UART_6 */
+}
+
+/* Initialize cache. */
+static void SOC_CacheInit(void)
+{
+	/* Enable System Bus Cache */
+	/* set command to invalidate all ways and write GO bit
+	 * to initiate command
+	 */
+	LMEM_PSCCR = LMEM_PSCCR_INVW1_MASK | LMEM_PSCCR_INVW0_MASK;
+	LMEM_PSCCR |= LMEM_PSCCR_GO_MASK;
+	/* Wait until the command completes */
+	while (LMEM_PSCCR & LMEM_PSCCR_GO_MASK)
+		;
+	/* Enable system bus cache, enable write buffer */
+	LMEM_PSCCR = (LMEM_PSCCR_ENWRBUF_MASK | LMEM_PSCCR_ENCACHE_MASK);
+	__ISB();
+
+	/* Enable Code Bus Cache */
+	/* set command to invalidate all ways and write GO bit
+	 * to initiate command
+	 */
+	LMEM_PCCCR = LMEM_PCCCR_INVW1_MASK | LMEM_PCCCR_INVW0_MASK;
+	LMEM_PCCCR |= LMEM_PCCCR_GO_MASK;
+	/* Wait until the command completes */
+	while (LMEM_PCCCR & LMEM_PCCCR_GO_MASK)
+		;
+	/* Enable code bus cache, enable write buffer */
+	LMEM_PCCCR = (LMEM_PCCCR_ENWRBUF_MASK | LMEM_PCCCR_ENCACHE_MASK);
+	__ISB();
+	__DSB();
+}
+
+/* Initialize clock. */
+static void SOC_ClockInit(void)
+{
+	/* OSC/PLL is already initialized by Cortex-A9 core */
+
+	/* Enable IP bridge and IO mux clock */
+	CCM_ControlGate(CCM, ccmCcgrGateIomuxIptClkIo, ccmClockNeededAll);
+	CCM_ControlGate(CCM, ccmCcgrGateIpmux1Clk, ccmClockNeededAll);
+	CCM_ControlGate(CCM, ccmCcgrGateIpmux2Clk, ccmClockNeededAll);
+	CCM_ControlGate(CCM, ccmCcgrGateIpmux3Clk, ccmClockNeededAll);
+
+#ifdef CONFIG_UART_IMX
+	/* Set UART clock is derived from OSC clock (24M) */
+	CCM_SetRootMux(CCM, ccmRootUartClkSel, ccmRootmuxUartClkOsc24m);
+
+	/* Configure UART divider */
+	CCM_SetRootDivider(CCM, ccmRootUartClkPodf, 0);
+
+	/* Enable UART clock */
+	CCM_ControlGate(CCM, ccmCcgrGateUartClk, ccmClockNeededAll);
+	CCM_ControlGate(CCM, ccmCcgrGateUartSerialClk, ccmClockNeededAll);
+#endif /* CONFIG_UART_IMX */
+}
+
+/**
+ *
+ * @brief Perform basic hardware initialization
+ *
+ * Initialize the interrupt controller device drivers.
+ * Also initialize the timer device driver, if required.
+ *
+ * @return 0
+ */
+static int mcimx6x_m4_init(struct device *arg)
```

Comment:
Should we have a similar soc initialization for nxp_mcimx7_init, like SOC_CacheInit()?

Reply:
I think this is optional. The FreeRTOS BSP initializes it when-in startup, see:
FreeRTOS_BSP_1.0.1_iMX7D/platform/devices/MCIMX7D/startup/system_MCIMX7D_M4.c, function SystemInit(). Or we could have a config item at the SoC level to enable/disable this.
We could also utilize LMEM_EnableSystemCache() in ext/hal/nxp/imx/drivers/lmem.c.